### PR TITLE
150 improve consistency

### LINKED ISSUE #150: Improve consistency in output
Right now the formatting of the `croncheck.php` is all over the place. The script is split into 'blocks', e.g. check if cron enabled, check adhoc/scheduled tasks, check check API. If any of these blocks reports a warning/error, it exits early.

This can lead to a situation where a long failing adhoc task masks any check API errors (which could be serious, e.g. objectfs connection failed)

The new format would look roughly like:

```
[Nagios status]: [Summary of ALL errors]
* [Error 1 title]
    [Error 1 details]
* [Error 2 title]
    [Error 2 details]
...
```
For example,
```
CRITICAL: Check API reported result errors:
* Dataflows
    HR Import: Run 16200 - aborted
    Details: https://test.com/report/status/index.php?detail=tool_dataflows_dataflow_runs
    https://test.com/admin/tool/dataflows/index.php
* Forced caching status
    Forced cache configuration is NOT active.
    Details: https://test.com/report/status/index.php?detail=tool_forcedcache_enabled
    https://test.com/admin/tool/forcedcache/index.php
* Task latency check
    Task \totara_core\task\tool_totara_sync_task has not run within the configured
    latency threshold: 2.
    Details: https://test.com/report/status/index.php?detail=tool_heartbeat_tasklatencycheck
* Task faildelay
    8 task(s) failing
    Fail delay: 1 day - tool_dataprivacy\task\process_data_request_task ID = 4
    Fail delay: 1 day - tool_dataprivacy\task\process_data_request_task ID = 5
    Fail delay: 1 day - tool_dataprivacy\task\process_data_request_task ID = 6

 (Checked Oct 10 13:00:50)
```

IMO the cleanest way to do this would be:

1. Fork this plugin for MOODLE_39_STABLE so we can assume check API exists and we don't have to care about legacy cron
2. Each of the errors are simply check results (scheduled and adhoc tasks are check classes, so we don't need to check these specifically)
3. Format these nicely and add the extra bit of logic to work out the nagios status and summary

Things to be mindful of though:
-  the only potential gotcha which that we'd have to double check (no pun) is that the check api was added and then various checks were added later than 3.9
- If the check API explodes (for e.g. bad plugin installed) it should report a critical with a clear message saying that this is masking real issues and should be fixed asap

### PR DESCRIPTION
**Pull request by matthewhilton**
Closes #150 

### Refactored `croncheck.php`

This page now ONLY displays the results of check API classes. This page works still either via web or CLI, however the configurable delays have been removed from this page, since those are handled by the check API class directly.

### Added new check class `failingtaskcheck.php`
While core does have an adhoc/scheduled task check, it has a few limitations:
1) It uses the task api which can explode if a bad task is installed masking failed tasks
2) It does not aggregate / deduplicate tasks leading to a big list if you have tons of a single task failing

So instead, I copied more or less the same code from https://github.com/catalyst/moodle-tool_heartbeat/pull/145 and put it inside a check class. I then suppress the core adhoc task checks from this page since all the information is now included nicely in the `failingtaskcheck.php` class.

### Example output: 

**No check failing**
```
OK: OK
(Checked Monday, 16 October 2023, 4:09 PM) 
```

**Single check failing (Shows error in top level)** (note for consistency, the message still shows in the list so it gets shown twice)
```
WARNING: Cron running: The admin/cron.php script has not been run for 15 mins 7 secs and should run every 1 min.
* WARNING Cron running: The admin/cron.php script has not been run for 15 mins 7 secs and should run every 1 min.
  15 mins 7 secs

(Checked Monday, 16 October 2023, 4:08 PM) 
```

**Multiple checks failing of different levels (shows summary of levels)**
```
CRITICAL: Multiple problems detected: 1 CRITICAL, 1 WARNING
* CRITICAL Failing tasks: 4 Moodle tasks reported errors, maximum faildelay > 600 mins
  SCHEDULED TASK: \core\task\h5p_get_content_types_task Delay: 86400
  ADHOC TASK: \tool_dataprivacy\task\process_data_request_task Delay: 86400 (13 duplicates!!!)
  ADHOC TASK: \tool_dataprivacy\task\initiate_data_request_task Delay: 86400
  ADHOC TASK: \core\task\build_installed_themes_task Delay: 86400

* WARNING Cron running: The admin/cron.php script has not been run for 3 mins 5 secs and should run every 1 min.
  3 mins 5 secs

(Checked Monday, 16 October 2023, 3:49 PM) 
```

** Debugging outputted and captured **
```
WARNING: Multiple problems detected: 2 WARNING
* WARNING Cron running: The admin/cron.php script has not been run for 13 mins 26 secs and should run every 1 min.
  13 mins 26 secs

* WARNING Output buffering: Unexpected output: testline 74 of /admin/tool/heartbeat/classes/checker.php: call to debugging()line 56 of /admin/tool/heartbeat/sitecheck.php: call to tool_heartbeat\checker::get_check_messages()
  

(Checked Monday, 16 October 2023, 4:06 PM) 
```

To Do:
- [x] Adhoc task + scheduled task - check output is useful
- [x] Debugging outputted - ensure captured by output buffer
- [x] Ci passes
- [x] Update README

### CODE REVIEW

File: sitecheck.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Check API Health Check
+ *
+ * @package    tool_heartbeat
+ * @copyright  2023 Matthew Hilton <matthewhilton@catalyst-au.net>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ *
+ * This can be run either as a web api, or on the CLI. When run on the
```
Comment (brendanheywood) [inline sitecheck.php:24]: copy pasta?

Comment (matthewhilton) [reply sitecheck.php]: yep

File: classes/checker.php [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        // Convert any line breaks to newlines.
+        $text = str_replace("<br />", "\n", $text);
+        $text = str_replace("<br>", "\n", $text);
+
+        // Strip tags.
+        $text = strip_tags($text);
+
+        // Clean any pipe characters from the $msg. This is because pipe characters
+        // separate Nagios performance data from log data.
+        $text = str_replace("|", "[pipe]", $text);
```
Comment (brendanheywood) [inline classes/checker.php:207]: lets swap this with a unicode vertical bar char

Comment (matthewhilton) [reply classes/checker.php]: Done

**Comment by matthewhilton** [comment]
## To do after merging:
- [ ] Update master branch README with supported branches table

**Pull request merged**